# Battle of the Neighborhoods - Coursera Capstone Project 


##### Author: Aaron E

## Introduction 

#### A local BBQ chain would like to expand to a new city. The prospect city is Toronto Canada. Toronto Canada is a rapidly developing city and provides various business oppourtunities. The city has become a major hub for tourist and businesses alike. Becuase of the growing compitetion in the area the stakeholders would like a detailed review of all the compitetion in the Toronto area and a detailed analysis of the surrounding areas.

#### We will be using Foursquare API to explore the restaurants in the Toronto area and to explore the most common venues while also using a list of zip codes from a wikipedia page to explore the neighborhoods.

#### from the data we will be able to assess the compitetion, and weather or not the area already has a sucessful BBQ restaurant. 

#### First lets explorea few choice boroughs in Toronoto, finding the most popular venues 

In [2]:
# Imported needed libraries
# importing old libraries (also used in the two first parts)
import pandas as pd
import numpy as np
from bs4 import BeautifulSoup
import requests
from bs4 import BeautifulSoup as bs
import json
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values
import requests # library to handle requests
import folium # map rendering library
print('Libraries imported.')

Libraries imported.


Read in the html as a Pandas Dataframe 

In [20]:
#get html from wiki page and create soup object
df = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0]
canada_df = df 
canada_df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Now we must clean up the data

Ignor the cells with bourough not assigned

In [21]:
# Ignore cells with Bourough not assigned
canada_df = canada_df[canada_df["Borough"] != "Not assigned"]

Group the data

In [23]:
df_toronto = canada_df.groupby(["Postal Code", "Borough"], as_index=False).agg(','.join)
df_toronto.head()

Postal Code      Borough                           Neighbourhood
0         M1B  Scarborough                          Malvern, Rouge
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

Replace mising values in Neighborhood with Borough if Neighborhood is not assigned

In [24]:
df_toronto['Neighbourhood'] = np.where(df_toronto['Neighbourhood'] == 'Not assigned',
                                              df_toronto['Borough'], df_toronto['Neighbourhood'])

Lets check the data shape

In [25]:
df_toronto.shape

(103, 3)

Lets get the geosatial data and add it to our dataframe

In [26]:
geospatial_url = "https://cocl.us/Geospatial_data"
geospatial_data = pd.read_csv(geospatial_url)

In [27]:
merged_data = pd.merge(df_toronto, geospatial_data, on='Postal Code')

In [31]:
df_toronto = merged_data

In [37]:
df_toronto.head(100)

Postal Code      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
..         ...          ...   
95         M9C    Etobicoke   
96         M9L   North York   
97         M9M   North York   
98         M9N         York   
99         M9P    Etobicoke   

                                        Neighbourhood   Latitude  Longitude  
0                                      Malvern, Rouge  43.806686 -79.194353  
1              Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497  
2                   Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                              Woburn  43.770992 -79.216917  
4                                           Cedarbrae  43.773136 -79.239476  
..                                                ...        ...        ...  
95  Eringate, Bloordale Gardens, Old Burnhamthorpe...  43.643515 -79.577201  
96                                      Humber Summit  43.756303 -79.565963  
97                                   Humberlea, Emery  43.724766 -79.532242  
98                                             Weston  43.706876 -79.518188  
99                                          Westmount  43.696319 -79.532242  

[100 rows x 5 columns]

## Lets see a map of Toronto with all the neighborhood 

In [33]:
address = "Toronto, Ontario"
geolocator = Nominatim(user_agent="ny_explorer", timeout=30)
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto 43.6534817, -79.3839347.


In [43]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, lng, label in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## This client is interested in a more populated part of the city, below we will anaylize Downtown Toronto.

In [36]:
# Credentials for Foursquare API
# @hidden_cell
CLIENT_ID = 'S0HWGHYZKXRMZUK0JN5FM1R2RDBN5XUYURRAEL01MIJCJYP5' # your Foursquare ID
CLIENT_SECRET = '4ILG5R1AX5H3JRJ3I5XGIOGB4EOZJBDRRSHQQKMH4NP344DE' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: S0HWGHYZKXRMZUK0JN5FM1R2RDBN5XUYURRAEL01MIJCJYP5
CLIENT_SECRET:4ILG5R1AX5H3JRJ3I5XGIOGB4EOZJBDRRSHQQKMH4NP344DE


In [58]:
down_town_data = df_toronto[df_toronto['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
james_town_data

Postal Code           Borough  \
0          M4W  Downtown Toronto   
1          M4X  Downtown Toronto   
2          M4Y  Downtown Toronto   
3          M5A  Downtown Toronto   
4          M5B  Downtown Toronto   
5          M5C  Downtown Toronto   
6          M5E  Downtown Toronto   
7          M5G  Downtown Toronto   
8          M5H  Downtown Toronto   
9          M5J  Downtown Toronto   
10         M5K  Downtown Toronto   
11         M5L  Downtown Toronto   
12         M5S  Downtown Toronto   
13         M5T  Downtown Toronto   
14         M5V  Downtown Toronto   
15         M5W  Downtown Toronto   
16         M5X  Downtown Toronto   
17         M6G  Downtown Toronto   
18         M7A  Downtown Toronto   

                                        Neighbourhood   Latitude  Longitude  
0                                            Rosedale  43.679563 -79.377529  
1                         St. James Town, Cabbagetown  43.667967 -79.367675  
2                                Church and Wellesley  43.665860 -79.383160  
3                           Regent Park, Harbourfront  43.654260 -79.360636  
4                            Garden District, Ryerson  43.657162 -79.378937  
5                                      St. James Town  43.651494 -79.375418  
6                                         Berczy Park  43.644771 -79.373306  
7                                  Central Bay Street  43.657952 -79.387383  
8                            Richmond, Adelaide, King  43.650571 -79.384568  
9   Harbourfront East, Union Station, Toronto Islands  43.640816 -79.381752  
10           Toronto Dominion Centre, Design Exchange  43.647177 -79.381576  
11                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
12                     University of Toronto, Harbord  43.662696 -79.400049  
13          Kensington Market, Chinatown, Grange Park  43.653206 -79.400049  
14  CN Tower, King and Spadina, Railway Lands, Har...  43.628947 -79.394420  
15                                     Stn A PO Boxes  43.646435 -79.374846  
16             First Canadian Place, Underground city  43.648429 -79.382280  
17                                           Christie  43.669542 -79.422564  
18        Queen's Park, Ontario Provincial Government  43.662301 -79.389494

In [59]:
james_town_data["Neighbourhood"].count()

19

# 19 neighborhoods in Downtown Toronoto

### Map of Downtown Toronto

In [44]:
address_scar = 'Downtown Toronto'
latitude_scar = 43.667967
longitude_scar = -79.3676675
print('The geograpical coordinate of Scarborough are {}, {}.'.format(latitude_scar, longitude_scar))

The geograpical coordinate of Scarborough are 43.667967, -79.3676675.


In [46]:
map_downtown = folium.Map(location=[latitude_scar, longitude_scar], zoom_start=12)

# add markers to map
for lat, lng, label in zip(down_town_data['Latitude'], down_town_data['Longitude'], down_town_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_downtown)  
    
map_downtown

### Top venues in Downtown Toronto

In [49]:
down_town_data.loc[1, "Neighbourhood"]

'St. James Town, Cabbagetown'

In [51]:
neighbourhood_lat = down_town_data.loc[1,"Latitude"] # Neighbourhood lat value
neighbourhood_long = down_town_data.loc[1, "Longitude"] # Neighbourhood long value
neighbourhood_name = down_town_data.loc[1, "Neighbourhood"] # neighbourhoods name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_lat, 
                                                               neighbourhood_long))

Latitude and longitude values of St. James Town, Cabbagetown are 43.667967, -79.3676753.


Setting a radius

In [118]:
LIMIT = 500
radius = 15000
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET,
                                                                                                                           latitude_scar, longitude_scar, 
                                                                                                                           VERSION, radius, LIMIT)


In [119]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5f21e8ffd46d721c653d9ccb'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Toronto',
  'headerFullLocation': 'Toronto',
  'headerLocationGranularity': 'city',
  'totalResults': 235,
  'suggestedBounds': {'ne': {'lat': 43.80296713500013,
    'lng': -79.18138497979078},
   'sw': {'lat': 43.532966864999864, 'lng': -79.55395002020921}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bb63b82941ad13a5e181fe3',
       'name': 'Riverdale Park West',
       'location': {'address': '500 Gerrard St.',
        'crossStreet': 'at River St.',
        'lat': 43.66604827676354,
        'lng': -79.36094117493359,
        'labeledLatLngs': [{'label': 'display'

In [120]:
# Function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [121]:
# structure and clean data into pandas dataframe 
venues = results['response']['groups'][0]['items']
    
nearby_venues = pd.json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues

name             categories        lat        lng
0     Riverdale Park West                   Park  43.666048 -79.360941
1     Riverdale Park East                   Park  43.669951 -79.355493
2     House on Parliament              Gastropub  43.663646 -79.367854
3            Zakkushi ざっ串    Japanese Restaurant  43.663618 -79.370670
4          Riverdale Farm                   Farm  43.666850 -79.361376
..                    ...                    ...        ...        ...
95           Lake Ontario                   Lake  43.638945 -79.379665
96    Harbourfront Centre  Performing Arts Venue  43.638556 -79.383190
97      The Moonbean Cafe                   Café  43.654147 -79.400182
98  Rodney's Oyster House     Seafood Restaurant  43.644975 -79.396587
99           The Sidekick             Comic Shop  43.664484 -79.325162

[100 rows x 4 columns]

In [122]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

100 venues were returned by Foursquare.


# 100 venues have been found in the surrounding areas of St.JamesTown in Downtown Toronto, we will analyize each one. 

In [123]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [124]:
downtown_neighborhood_venues = getNearbyVenues(names=down_town_data['Neighbourhood'],
                                   latitudes=down_town_data['Latitude'],
                                   longitudes=down_town_data['Longitude']
                                  )

Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Queen's Park, Ontario Provincial Government


In [125]:
downtown_neighborhood_venues.head(15)

Neighborhood  Neighborhood Latitude  \
0                      Rosedale              43.679563   
1                      Rosedale              43.679563   
2                      Rosedale              43.679563   
3                      Rosedale              43.679563   
4   St. James Town, Cabbagetown              43.667967   
5   St. James Town, Cabbagetown              43.667967   
6   St. James Town, Cabbagetown              43.667967   
7   St. James Town, Cabbagetown              43.667967   
8   St. James Town, Cabbagetown              43.667967   
9   St. James Town, Cabbagetown              43.667967   
10  St. James Town, Cabbagetown              43.667967   
11  St. James Town, Cabbagetown              43.667967   
12  St. James Town, Cabbagetown              43.667967   
13  St. James Town, Cabbagetown              43.667967   
14  St. James Town, Cabbagetown              43.667967   

    Neighborhood Longitude                       Venue  Venue Latitude  \
0               -79.377529               Rosedale Park       43.682328   
1               -79.377529                Whitney Park       43.682036   
2               -79.377529        Alex Murray Parkette       43.678300   
3               -79.377529              Milkman's Lane       43.676352   
4               -79.367675                 Cranberries       43.667843   
5               -79.367675              Kingyo Toronto       43.665895   
6               -79.367675      Butter Chicken Factory       43.667072   
7               -79.367675                    F'Amelia       43.667536   
8               -79.367675                  Murgatroid       43.667381   
9               -79.367675    Merryberry Cafe + Bistro       43.666630   
10              -79.367675            Cabbagetown Brew       43.666923   
11              -79.367675      Absolute Bakery & Café       43.667469   
12              -79.367675    Fair Trade Jewellery Co.       43.665348   
13              -79.367675       Toronto Dance Theatre       43.666232   
14              -79.367675  St. Jamestown Delicatessen       43.665811   

    Venue Longitude         Venue Category  
0        -79.378934             Playground  
1        -79.373788                   Park  
2        -79.382773                   Park  
3        -79.373842                  Trail  
4        -79.369407                  Diner  
5        -79.368415    Japanese Restaurant  
6        -79.369184      Indian Restaurant  
7        -79.368613     Italian Restaurant  
8        -79.369311             Restaurant  
9        -79.368792                   Café  
10       -79.369289                   Café  
11       -79.369277                 Bakery  
12       -79.368362          Jewelry Store  
13       -79.367075  General Entertainment  
14       -79.368648                Butcher

In [85]:
downtown_neighborhood_venues.groupby("Neighborhood").count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            58   
CN Tower, King and Spadina, Railway Lands, Harb...                     14   
Central Bay Street                                                     62   
Christie                                                               17   
Church and Wellesley                                                   76   
Commerce Court, Victoria Hotel                                        100   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
Kensington Market, Chinatown, Grange Park                              65   
Queen's Park, Ontario Provincial Government                            31   
Regent Park, Harbourfront                                              46   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
St. James Town                                                         85   
St. James Town, Cabbagetown                                            45   
Stn A PO Boxes                                                         98   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             58   
CN Tower, King and Spadina, Railway Lands, Harb...                      14   
Central Bay Street                                                      62   
Christie                                                                17   
Church and Wellesley                                                    76   
Commerce Court, Victoria Hotel                                         100   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               65   
Queen's Park, Ontario Provincial Government                             31   
Regent Park, Harbourfront                                               46   
Richmond, Adelaide, King                                               100   
Rosedale                                                                 4   
St. James Town                                                          85   
St. James Town, Cabbagetown                                             45   
Stn A PO Boxes                                                          98   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          34   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            58              58   
CN Tower, King and Spadina, Railway Lands, Harb...     14              14   
Central Bay Street                                     62              62   
Christie                                               17              17   
Church and Wellesley                                   76              76   
Commerce Court, Victoria Hotel                        100             100   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100             100   
Harbourfront East, Union 

In [126]:
print('There are {} uniques categories.'.format(len(downtown_neighborhood_venues['Venue Category'].unique())))

There are 209 uniques categories.


# Count of how many venues are in each neighborhood; there are 209 unique categories in Downtown Toronto

# Next we will see the type of venues in each neighborhood

In [127]:
# one hot encoding
downtown_onehot = pd.get_dummies(downtown_neighborhood_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
downtown_onehot['Neighbourhood'] = downtown_neighborhood_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [downtown_onehot.columns[-1]] + list(downtown_onehot.columns[:-1])
downtown_onehot = downtown_onehot[fixed_columns]

downtown_onehot.head()

Neighbourhood  Afghan Restaurant  Airport  \
0                     Rosedale                  0        0   
1                     Rosedale                  0        0   
2                     Rosedale                  0        0   
3                     Rosedale                  0        0   
4  St. James Town, Cabbagetown                  0        0   

   Airport Food Court  Airport Lounge  Airport Service  Airport Terminal  \
0                   0               0                0                 0   
1                   0               0                0                 0   
2                   0               0                0                 0   
3                   0               0                0                 0   
4                   0               0                0                 0   

   American Restaurant  Antique Shop  Aquarium  ...  Toy / Game Store  Trail  \
0                    0             0         0  ...                 0      0   
1                    0             0         0  ...                 0      0   
2                    0             0         0  ...                 0      0   
3                    0             0         0  ...                 0      1   
4                    0             0         0  ...                 0      0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0              0                              0                 0   
1              0                              0                 0   
2              0                              0                 0   
3              0                              0                 0   
4              0                              0                 0   

   Vietnamese Restaurant  Wine Bar  Wine Shop  Women's Store  Yoga Studio  
0                      0         0          0              0            0  
1                      0         0          0              0            0  
2                      0         0          0              0            0  
3                      0         0          0              0            0  
4                      0         0          0              0            0  

[5 rows x 210 columns]

Grouped neighborhoods by mean of frewuency of ocurrence of each category 

In [128]:
downtown_grouped = downtown_onehot.groupby('Neighborhood').mean().reset_index()
downtown_grouped.head()

Neighborhood  Afghan Restaurant   Airport  Airport Food Court  \
0             0           0.000811  0.000811            0.000811   
1             1           0.000000  0.000000            0.000000   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0        0.001622         0.001622          0.001622              0.01541   
1        0.000000         0.000000          0.000000              0.00000   

   Antique Shop  Aquarium  ...  Toy / Game Store     Trail  Train Station  \
0      0.001622  0.004055  ...          0.000811  0.000811       0.002433   
1      0.000000  0.000000  ...          0.000000  0.000000       0.000000   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                       0.011354          0.000811               0.003244   
1                       0.000000          0.000000               0.000000   

   Wine Bar  Wine Shop  Women's Store  Yoga Studio  
0  0.006488   0.000811       0.000811     0.005677  
1  0.000000   0.000000       0.000000     0.000000  

[2 rows x 209 columns]

# Lets simplify and find the top 10 venues 

In [129]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [130]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = downtown_grouped['Neighborhood']

for ind in np.arange(downtown_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(downtown_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0             0           Coffee Shop                  Café   
1             1           Yoga Studio         Women's Store   

  3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
0            Restaurant                        Hotel   Japanese Restaurant   
1     Electronics Store  Eastern European Restaurant   Dumpling Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0    Italian Restaurant                  Park    Seafood Restaurant   
1            Donut Shop      Doner Restaurant               Dog Run   

  9th Most Common Venue 10th Most Common Venue  
0                   Gym                 Bakery  
1   Distribution Center         Discount Store

# The most popular venues in the neighborhood of JamesTown, Rosedale, and Cabbage Town in the borough of Downtown Toronto is showed above.

# The most popular venue is a coffee shop in Rosedale, and a Yoga Studio in Jamestown/Cabbage Town.

# Lets now explore another area

In [137]:
scarborough_data = df_toronto[df_toronto['Borough'] == 'Scarborough'].reset_index(drop=True)
scarborough_data.head()

Postal Code      Borough                           Neighbourhood   Latitude  \
0         M1B  Scarborough                          Malvern, Rouge  43.806686   
1         M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

### Map of Scarborough

In [138]:
address_scar = 'Scarborough,Toronto'
latitude_scar = 43.773077
longitude_scar = -79.257774
print('The geograpical coordinate of Scarborough are {}, {}.'.format(latitude_scar, longitude_scar))

The geograpical coordinate of Scarborough are 43.773077, -79.257774.


In [139]:
map_scarborough = folium.Map(location=[latitude_scar, longitude_scar], zoom_start=12)

# add markers to map
for lat, lng, label in zip(scarborough_data['Latitude'], scarborough_data['Longitude'], scarborough_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_scarborough)  
    
map_scarborough

### Top venues in the neighborhood "Scarborough Village"

In [140]:
scarborough_data.loc[5, "Neighbourhood"]

'Scarborough Village'

In [141]:
neighbourhood_lat = scarborough_data.loc[5,"Latitude"] # Neighbourhood lat value
neighbourhood_long = scarborough_data.loc[5, "Longitude"] # Neighbourhood long value
neighbourhood_name = scarborough_data.loc[5, "Neighbourhood"] # neighbourhoods name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_lat, 
                                                               neighbourhood_long))

Latitude and longitude values of Scarborough Village are 43.7447342, -79.23947609999999.


In [153]:
LIMIT = 500
radius = 15000
url = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&v={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET,
                                                                                                                           latitude_scar, longitude_scar, 
                                                                                                                           VERSION, radius, LIMIT)


In [154]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5f21eb5272cee757fc3ef7de'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Toronto',
  'headerFullLocation': 'Toronto',
  'headerLocationGranularity': 'city',
  'totalResults': 233,
  'suggestedBounds': {'ne': {'lat': 43.908077135000134,
    'lng': -79.07116438641437},
   'sw': {'lat': 43.63807686499987, 'lng': -79.44438361358563}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4c5c9bee94fd0f47cb30c945',
       'name': 'The Roti Hut',
       'location': {'address': '351 Pitfield Rd',
        'crossStreet': 'McCowan Rd',
        'lat': 43.78727725968595,
        'lng': -79.25872401779647,
        'labeledLatLngs': [{'label': 'display',
       

In [148]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [157]:
# structure and clean data into pandas dataframe 
venues = results['response']['groups'][0]['items']
    
nearby_venues = pd.json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat  \
0                      The Roti Hut  Caribbean Restaurant  43.787277   
1                       Mona's Roti  Caribbean Restaurant  43.791613   
2                   CANBE Foods Inc     Indian Restaurant  43.773546   
3  The Real McCoy Burgers And Pizza          Burger Joint  43.774081   
4             Thomson Memorial Park                  Park  43.758891   

         lng  
0 -79.258724  
1 -79.251015  
2 -79.246082  
3 -79.230496  
4 -79.254500

In [158]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

100 venues were returned by Foursquare.


In [161]:
scarborough_nieghnorhood_venues = getNearbyVenues(names=scarborough_data['Neighbourhood'],
                                   latitudes=scarborough_data['Latitude'],
                                   longitudes=scarborough_data['Longitude']
                                  )

Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge


In [162]:
scarborough_nieghnorhood_venues.head()

Neighborhood  Neighborhood Latitude  \
0                          Malvern, Rouge              43.806686   
1  Rouge Hill, Port Union, Highland Creek              43.784535   
2  Rouge Hill, Port Union, Highland Creek              43.784535   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                        Venue  Venue Latitude  \
0              -79.194353                      Wendy’s       43.807448   
1              -79.160497  Great Shine Window Cleaning       43.783145   
2              -79.160497        Royal Canadian Legion       43.782533   
3              -79.188711               RBC Royal Bank       43.766790   
4              -79.188711            G & G Electronics       43.765309   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.157431          Home Service  
2       -79.163085                   Bar  
3       -79.191151                  Bank  
4       -79.191537     Electronics Store

In [163]:
scarborough_nieghnorhood_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               4   
Birch Cliff, Cliffside West                                             4   
Cedarbrae                                                               8   
Clarks Corners, Tam O'Shanter, Sullivan                                12   
Cliffside, Cliffcrest, Scarborough Village West                         2   
Dorset Park, Wexford Heights, Scarborough Town ...                      5   
Golden Mile, Clairlea, Oakridge                                         8   
Guildwood, Morningside, West Hill                                       8   
Kennedy Park, Ionview, East Birchmount Park                             5   
Malvern, Rouge                                                          1   
Milliken, Agincourt North, Steeles East, L'Amor...                      3   
Rouge Hill, Port Union, Highland Creek                                  2   
Scarborough Village                                                     2   
Steeles West, L'Amoreaux West                                          15   
Wexford, Maryvale                                                       7   
Woburn                                                                  4   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Agincourt                                                                4   
Birch Cliff, Cliffside West                                              4   
Cedarbrae                                                                8   
Clarks Corners, Tam O'Shanter, Sullivan                                 12   
Cliffside, Cliffcrest, Scarborough Village West                          2   
Dorset Park, Wexford Heights, Scarborough Town ...                       5   
Golden Mile, Clairlea, Oakridge                                          8   
Guildwood, Morningside, West Hill                                        8   
Kennedy Park, Ionview, East Birchmount Park                              5   
Malvern, Rouge                                                           1   
Milliken, Agincourt North, Steeles East, L'Amor...                       3   
Rouge Hill, Port Union, Highland Creek                                   2   
Scarborough Village                                                      2   
Steeles West, L'Amoreaux West                                           15   
Wexford, Maryvale                                                        7   
Woburn                                                                   4   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Agincourt                                               4               4   
Birch Cliff, Cliffside West                             4               4   
Cedarbrae                                               8               8   
Clarks Corners, Tam O'Shanter, Sullivan                12              12   
Cliffside, Cliffcrest, Scarborough Village West         2               2   
Dorset Park, Wexford Heights, Scarborough Town ...      5               5   
Golden Mile, Clairlea, Oakridge                         8               8   
Guildwood, Morningside, West Hill                       8               8   
Kennedy Park, Ionview, East Birchmount Park             5               5   
Malvern, Rouge                                          1               1   
Milliken, Agincourt North, Steeles East, L'Amor...      3               3   
Rouge Hill, Port Union, Highland Creek                  2               2   
Scarborough Village                                     2               2   
Steeles West, L'Amoreaux West                          15              15   
Wexford, Maryvale           

In [164]:
print('There are {} uniques categories.'.format(len(scarborough_nieghnorhood_venues['Venue Category'].unique())))

There are 56 uniques categories.


# There are 56 unique categories in the borough of Scarborough

In [165]:
# one hot encoding
scarborough_onehot = pd.get_dummies(scarborough_nieghnorhood_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
scarborough_onehot['Neighborhood'] = scarborough_nieghnorhood_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [scarborough_onehot.columns[-1]] + list(scarborough_onehot.columns[:-1])
scarborough_onehot = scarborough_onehot[fixed_columns]

scarborough_onehot.head()

Neighborhood  American Restaurant  \
0                          Malvern, Rouge                    0   
1  Rouge Hill, Port Union, Highland Creek                    0   
2  Rouge Hill, Port Union, Highland Creek                    0   
3       Guildwood, Morningside, West Hill                    0   
4       Guildwood, Morningside, West Hill                    0   

   Athletics & Sports  Auto Garage  Bakery  Bank  Bar  Breakfast Spot  \
0                   0            0       0     0    0               0   
1                   0            0       0     0    0               0   
2                   0            0       0     0    1               0   
3                   0            0       0     1    0               0   
4                   0            0       0     0    0               0   

   Bus Line  Café  ...  Restaurant  Sandwich Place  Sculpture Garden  \
0         0     0  ...           0               0                 0   
1         0     0  ...           0               0                 0   
2         0     0  ...           0               0                 0   
3         0     0  ...           0               0                 0   
4         0     0  ...           0               0                 0   

   Shopping Mall  Skating Rink  Soccer Field  Thai Restaurant  \
0              0             0             0                0   
1              0             0             0                0   
2              0             0             0                0   
3              0             0             0                0   
4              0             0             0                0   

   Thrift / Vintage Store  Vietnamese Restaurant  Women's Store  
0                       0                      0              0  
1                       0                      0              0  
2                       0                      0              0  
3                       0                      0              0  
4                       0                      0              0  

[5 rows x 57 columns]

grouped neighborhoods by mean of frequency of occurrence of each category

In [166]:
scarborough_grouped = scarborough_onehot.groupby('Neighborhood').mean().reset_index()
scarborough_grouped.head(7)

Neighborhood  American Restaurant  \
0                                          Agincourt                  0.0   
1                        Birch Cliff, Cliffside West                  0.0   
2                                          Cedarbrae                  0.0   
3            Clarks Corners, Tam O'Shanter, Sullivan                  0.0   
4    Cliffside, Cliffcrest, Scarborough Village West                  0.5   
5  Dorset Park, Wexford Heights, Scarborough Town...                  0.0   
6                    Golden Mile, Clairlea, Oakridge                  0.0   

   Athletics & Sports  Auto Garage  Bakery      Bank  Bar  Breakfast Spot  \
0               0.000          0.0   0.000  0.000000  0.0            0.25   
1               0.000          0.0   0.000  0.000000  0.0            0.00   
2               0.125          0.0   0.125  0.125000  0.0            0.00   
3               0.000          0.0   0.000  0.083333  0.0            0.00   
4               0.000          0.0   0.000  0.000000  0.0            0.00   
5               0.000          0.0   0.000  0.000000  0.0            0.00   
6               0.000          0.0   0.250  0.000000  0.0            0.00   

   Bus Line  Café  ...  Restaurant  Sandwich Place  Sculpture Garden  \
0     0.000  0.00  ...         0.0             0.0               0.0   
1     0.000  0.25  ...         0.0             0.0               0.0   
2     0.000  0.00  ...         0.0             0.0               0.0   
3     0.000  0.00  ...         0.0             0.0               0.0   
4     0.000  0.00  ...         0.0             0.0               0.0   
5     0.000  0.00  ...         0.0             0.0               0.0   
6     0.125  0.00  ...         0.0             0.0               0.0   

   Shopping Mall  Skating Rink  Soccer Field  Thai Restaurant  \
0            0.0          0.00         0.000         0.000000   
1            0.0          0.25         0.000         0.000000   
2            0.0          0.00         0.000         0.125000   
3            0.0          0.00         0.000         0.083333   
4            0.0          0.00         0.000         0.000000   
5            0.0          0.00         0.000         0.000000   
6            0.0          0.00         0.125         0.000000   

   Thrift / Vintage Store  Vietnamese Restaurant  Women's Store  
0                     0.0                    0.0            0.0  
1                     0.0                    0.0            0.0  
2                     0.0                    0.0            0.0  
3                     0.0                    0.0            0.0  
4                     0.0                    0.0            0.0  
5                     0.0                    0.2            0.0  
6                     0.0                    0.0            0.0  

[7 rows x 57 columns]

### top 10 venues per neighboring in Scarborough

In [167]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = scarborough_grouped['Neighborhood']

for ind in np.arange(scarborough_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(scarborough_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted

Neighborhood  \
0                                           Agincourt   
1                         Birch Cliff, Cliffside West   
2                                           Cedarbrae   
3             Clarks Corners, Tam O'Shanter, Sullivan   
4     Cliffside, Cliffcrest, Scarborough Village West   
5   Dorset Park, Wexford Heights, Scarborough Town...   
6                     Golden Mile, Clairlea, Oakridge   
7                   Guildwood, Morningside, West Hill   
8         Kennedy Park, Ionview, East Birchmount Park   
9                                      Malvern, Rouge   
10  Milliken, Agincourt North, Steeles East, L'Amo...   
11             Rouge Hill, Port Union, Highland Creek   
12                                Scarborough Village   
13                      Steeles West, L'Amoreaux West   
14                                  Wexford, Maryvale   
15                                             Woburn   

        1st Most Common Venue 2nd Most Common Venue  3rd Most Common Venue  \
0   Latin American Restaurant        Breakfast Spot                 Lounge   
1             College Stadium          Skating Rink  General Entertainment   
2          Athletics & Sports       Thai Restaurant       Hakka Restaurant   
3                 Pizza Place              Pharmacy   Fast Food Restaurant   
4         American Restaurant                 Motel           Home Service   
5           Indian Restaurant             Pet Store     Chinese Restaurant   
6                      Bakery        Ice Cream Shop           Intersection   
7          Mexican Restaurant          Intersection                   Bank   
8           Convenience Store            Hobby Shop         Discount Store   
9        Fast Food Restaurant         Women's Store  Vietnamese Restaurant   
10           Sculpture Garden            Playground                   Park   
11               Home Service                   Bar      Convenience Store   
12              Women's Store            Playground        College Stadium   
13         Chinese Restaurant  Fast Food Restaurant         Breakfast Spot   
14             Sandwich Place         Shopping Mall    Fried Chicken Joint   
15                Coffee Shop     Convenience Store      Korean Restaurant   

    4th Most Common Venue      5th Most Common Venue 6th Most Common Venue  \
0          Clothing Store              Women's Store      Department Store   
1                    Café              Women's Store      Hakka Restaurant   
2                  Bakery                       Bank           Gas Station   
3            Noodle House            Thai Restaurant           Gas Station   
4        Hakka Restaurant                        Gym         Grocery Store   
5   Vietnamese Restaurant                     Bakery      Department Store   
6           Metro Station                   Bus Line                  Park   
7          Medical Center             Breakfast Spot     Electronics Store   
8        Department Store                Coffee Shop         Women's Store   
9              Hobby Shop           Hakka Restaurant                   Gym   
10          Women's Store                Coffee Shop                   Gym   
11             Hobby Shop           Hakka Restaurant                   Gym   
12       Hakka Restaurant                        Gym         Grocery Store   
13          Grocery Store          Indian Restaurant        Discount Store   
14         Breakfast Spot  Middle Eastern Restaurant                Bakery   
15          Women's Store           Hakka Restaurant                   Gym   

    7th Most Common Venue  8th Most Common Venue  9th Most Common Venue  \
0        Hakka Restaurant                    Gym          Grocery Store   
1                     Gym          Grocery Store            Gas Station   
2    Caribbean Restaurant    Fried Chicken Joint          Women's Store   
3                    Bank     Italian Restaurant    Fried Chicken Joint   
4   General Entertainment            Gas S

# Cluster Neighborhoods in Scarborough using k-means

In [170]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

scar_data = scarborough_data.drop(14)
# set number of clusters
kclusters = 5

scarborough_grouped_clustering = scarborough_grouped.drop('Neighborhood', 1)


# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(scarborough_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 1, 1, 1, 0, 1, 1, 1, 3, 2])

In [186]:
scarborough_merged = scar_data

# add clustering labels
scarborough_merged['Cluster Labels'] = kmeans.labels_

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
scarborough_merged = scarborough_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')

scarborough_merged.head(15)

Postal Code      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
5          M1J  Scarborough   
6          M1K  Scarborough   
7          M1L  Scarborough   
8          M1M  Scarborough   
9          M1N  Scarborough   
10         M1P  Scarborough   
11         M1R  Scarborough   
12         M1S  Scarborough   
13         M1T  Scarborough   
15         M1W  Scarborough   

                                        Neighbourhood   Latitude  Longitude  \
0                                      Malvern, Rouge  43.806686 -79.194353   
1              Rouge Hill, Port Union, Highland Creek  43.784535 -79.160497   
2                   Guildwood, Morningside, West Hill  43.763573 -79.188711   
3                                              Woburn  43.770992 -79.216917   
4                                           Cedarbrae  43.773136 -79.239476   
5                                 Scarborough Village  43.744734 -79.239476   
6         Kennedy Park, Ionview, East Birchmount Park  43.727929 -79.262029   
7                     Golden Mile, Clairlea, Oakridge  43.711112 -79.284577   
8     Cliffside, Cliffcrest, Scarborough Village West  43.716316 -79.239476   
9                         Birch Cliff, Cliffside West  43.692657 -79.264848   
10  Dorset Park, Wexford Heights, Scarborough Town...  43.757410 -79.273304   
11                                  Wexford, Maryvale  43.750072 -79.295849   
12                                          Agincourt  43.794200 -79.262029   
13            Clarks Corners, Tam O'Shanter, Sullivan  43.781638 -79.304302   
15                      Steeles West, L'Amoreaux West  43.799525 -79.318389   

    Cluster Labels      1st Most Common Venue 2nd Most Common Venue  \
0                1       Fast Food Restaurant         Women's Store   
1                1               Home Service                   Bar   
2                1         Mexican Restaurant          Intersection   
3                1                Coffee Shop     Convenience Store   
4                0         Athletics & Sports       Thai Restaurant   
5                1              Women's Store            Playground   
6                1          Convenience Store            Hobby Shop   
7                1                     Bakery        Ice Cream Shop   
8                3        American Restaurant                 Motel   
9                2            College Stadium          Skating Rink   
10               1          Indian Restaurant             Pet Store   
11               1             Sandwich Place         Shopping Mall   
12               4  Latin American Restaurant        Breakfast Spot   
13               1                Pizza Place              Pharmacy   
15               1         Chinese Restaurant  Fast Food Restaurant   

    3rd Most Common Venue  4th Most Common Venue      5th Most Common Venue  \
0   Vietnamese Restaurant             Hobby Shop           Hakka Restaurant   
1       Convenience Store             Hobby Shop           Hakka Restaurant   
2                    Bank         Medical Center             Breakfast Spot   
3       Korean Restaurant          Women's Store           Hakka Restaurant   
4        Hakka Restaurant                 Bakery                       Bank   
5         College Stadium       Hakka Restaurant                        Gym   
6          Discount Store       Department Store                Coffee Shop   
7            Intersection          Metro Station                   Bus Line   
8            Home Service       Hakka Restaurant                        Gym   
9   General Entertainment                   Café              Women's Store   
10     Chinese Restaurant  Vietnamese Restaurant                     Bakery   
11    Fried Chicken Joint         Breakfast Spot  Middle Eastern Restaurant   
12                 Lounge         Clothing Store              Women's Store  

In [175]:
# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# create map
map_clusters = folium.Map(location = [latitude_scar, longitude_scar], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(scarborough_merged['Latitude'], scarborough_merged['Longitude'], scarborough_merged['Neighbourhood'], scarborough_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

# Examing the clusters 

Cluster 1

In [178]:
scarborough_merged.loc[scarborough_merged['Cluster Labels'] == 0, scarborough_merged.columns[[1] + list(range(5, scarborough_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
4  Scarborough               0    Athletics & Sports       Thai Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4      Hakka Restaurant                Bakery                  Bank   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4           Gas Station  Caribbean Restaurant   Fried Chicken Joint   

  9th Most Common Venue 10th Most Common Venue  
4         Women's Store       Department Store

Cluster 2

In [179]:
scarborough_merged.loc[scarborough_merged['Cluster Labels'] == 1, scarborough_merged.columns[[1] + list(range(5, scarborough_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0   Scarborough               1  Fast Food Restaurant         Women's Store   
1   Scarborough               1          Home Service                   Bar   
2   Scarborough               1    Mexican Restaurant          Intersection   
3   Scarborough               1           Coffee Shop     Convenience Store   
5   Scarborough               1         Women's Store            Playground   
6   Scarborough               1     Convenience Store            Hobby Shop   
7   Scarborough               1                Bakery        Ice Cream Shop   
10  Scarborough               1     Indian Restaurant             Pet Store   
11  Scarborough               1        Sandwich Place         Shopping Mall   
13  Scarborough               1           Pizza Place              Pharmacy   
15  Scarborough               1    Chinese Restaurant  Fast Food Restaurant   

    3rd Most Common Venue  4th Most Common Venue      5th Most Common Venue  \
0   Vietnamese Restaurant             Hobby Shop           Hakka Restaurant   
1       Convenience Store             Hobby Shop           Hakka Restaurant   
2                    Bank         Medical Center             Breakfast Spot   
3       Korean Restaurant          Women's Store           Hakka Restaurant   
5         College Stadium       Hakka Restaurant                        Gym   
6          Discount Store       Department Store                Coffee Shop   
7            Intersection          Metro Station                   Bus Line   
10     Chinese Restaurant  Vietnamese Restaurant                     Bakery   
11    Fried Chicken Joint         Breakfast Spot  Middle Eastern Restaurant   
13   Fast Food Restaurant           Noodle House            Thai Restaurant   
15         Breakfast Spot          Grocery Store          Indian Restaurant   

   6th Most Common Venue  7th Most Common Venue  8th Most Common Venue  \
0                    Gym          Grocery Store  General Entertainment   
1                    Gym          Grocery Store  General Entertainment   
2      Electronics Store             Restaurant    Rental Car Location   
3                    Gym          Grocery Store  General Entertainment   
5          Grocery Store  General Entertainment            Gas Station   
6          Women's Store       Hakka Restaurant                    Gym   
7                   Park           Soccer Field                   Bank   
10      Department Store     Athletics & Sports       Hakka Restaurant   
11                Bakery            Auto Garage          Grocery Store   
13           Gas Station                   Bank     Italian Restaurant   
15        Discount Store            Coffee Shop               Pharmacy   

    9th Most Common Venue 10th Most Common Venue  
0             Gas Station    Fried Chicken Joint  
1             Gas Station    Fried Chicken Joint  
2           Women's Store   Fast Food Restaurant  
3             Gas Station    Fried Chicken Joint  
5     Fried Chicken Joint   Fast Food Restaurant  
6           Grocery Store  General Entertainment  
7          Discount Store                    Gym  
10                    Gym          Grocery Store  
11  General Entertainment      Convenience Store  
13    Fried Chicken Joint     Chinese Restaurant  
15            Pizza Place                    Gym

Cluster 3 

In [180]:
scarborough_merged.loc[scarborough_merged['Cluster Labels'] == 2, scarborough_merged.columns[[1] + list(range(5, scarborough_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
9  Scarborough               2       College Stadium          Skating Rink   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
9  General Entertainment                  Café         Women's Store   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
9      Hakka Restaurant                   Gym         Grocery Store   

  9th Most Common Venue 10th Most Common Venue  
9           Gas Station    Fried Chicken Joint

Cluster 4

In [181]:
scarborough_merged.loc[scarborough_merged['Cluster Labels'] == 3, scarborough_merged.columns[[1] + list(range(5, scarborough_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
8   Scarborough               3   American Restaurant                 Motel   
16  Scarborough               3                   NaN                   NaN   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
8           Home Service      Hakka Restaurant                   Gym   
16                   NaN                   NaN                   NaN   

   6th Most Common Venue  7th Most Common Venue 8th Most Common Venue  \
8          Grocery Store  General Entertainment           Gas Station   
16                   NaN                    NaN                   NaN   

   9th Most Common Venue 10th Most Common Venue  
8    Fried Chicken Joint   Fast Food Restaurant  
16                   NaN                    NaN

Cluster 5

In [183]:
scarborough_merged.loc[scarborough_merged['Cluster Labels'] == 4, scarborough_merged.columns[[1] + list(range(5, scarborough_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
12  Scarborough               4  Latin American Restaurant   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
12        Breakfast Spot                Lounge        Clothing Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
12         Women's Store      Department Store      Hakka Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
12                   Gym         Grocery Store  General Entertainment